File: solutions/kotlin/resistor-color-trio/2/src/main/kotlin/ResistorColorTrio.kt
import Unit
import java.lang.IllegalArgumentException

object ResistorColorTrio {

    fun text(vararg input: Color): String {
        require(input[0].ordinal != 0)
        var zeros = input[2].ordinal
        val begin: String

        if (input[1].ordinal == 0) {
            begin = (input[0].ordinal).toString()
            zeros += 1
        } else {
            begin = (input[0].ordinal * 10 + input[1].ordinal).toString()
        }

        val end = when {
            zeros < 3 -> "${"0".repeat(zeros)} ${Unit.OHMS.name.lowercase()}"
            zeros < 6 -> "${"0".repeat(zeros - 3)} ${Unit.KILOOHMS.name.lowercase()}"
            zeros < 9 -> "${"0".repeat(zeros - 6)} ${Unit.MEGAOHMS.name.lowercase()}"
            zeros < 12 -> "${"0".repeat(zeros - 9)} ${Unit.GIGAOHMS.name.lowercase()}"
            else -> "Oh, this is too much..."
        }


        return begin + end
    }
}

File: solutions/go/resistor-color-trio/1/resistor_color_trio.go
package resistorcolortrio

import (
	"strconv"
	"strings"
)

// Label describes the resistance value given the colors of a resistor.
// The label is a string with a resistance value with an unit appended
// (e.g. "33 ohms", "470 kiloohms").

var mapping = map[string]int{
	"black":  0,
	"brown":  1,
	"red":    2,
	"orange": 3,
	"yellow": 4,
	"green":  5,
	"blue":   6,
	"violet": 7,
	"grey":   8,
	"white":  9,
}

func Label(colors []string) string {
	zeros := mapping[colors[2]]
	var begin string
	var end string

	firstDigit := mapping[colors[0]]
	secondDigit := mapping[colors[1]]
	if secondDigit == 0 {
		begin = strconv.Itoa(firstDigit)
		zeros += 1
	} else {
		begin = strconv.Itoa(firstDigit*10 + secondDigit)
	}

	switch {
	case zeros < 3:
		end = strings.Repeat("0", zeros) + " ohms"
	case zeros < 6:
		end = strings.Repeat("0", zeros-3) + " kiloohms"
	case zeros < 9:
		end = strings.Repeat("0", zeros-6) + " megaohms"
	case zeros < 12:
		end = strings.Repeat("0", zeros-9) + " gigaohms"
	default:
		panic("this is too much")
	}

	resistance := begin + end
	if strings.HasPrefix(resistance, "0") {
		return resistance[1:]
	}
	return resistance
}
